File: frontend/lib/contracts/ContentRewards.ts
import { createClient } from "genlayer-js";
import { studionet } from "genlayer-js/chains";
import type {
  Contest,
  Submission,
  UserSubmission,
  SubmitResult,
  TransactionReceipt,
} from "./types";

/**
 * Helper to convert Map to plain object recursively
 */
function mapToObject(data: any): any {
  if (data instanceof Map) {
    const obj: Record<string, any> = {};
    for (const [key, value] of data.entries()) {
      obj[key] = mapToObject(value);
    }
    return obj;
  }
  if (Array.isArray(data)) {
    return data.map(mapToObject);
  }
  return data;
}

/**
 * ContentRewards contract class for interacting with the GenLayer Content Rewards contract
 */
class ContentRewards {
  private contractAddress: `0x${string}`;
  private client: ReturnType<typeof createClient>;

  constructor(
    contractAddress: string,
    address?: string | null,
    studioUrl?: string
  ) {
    this.contractAddress = contractAddress as `0x${string}`;

    const config: any = {
      chain: studionet,
    };

    if (address) {
      config.account = address as `0x${string}`;
    }

    if (studioUrl) {
      config.endpoint = studioUrl;
    }

    this.client = createClient(config);
  }

  /**
   * Update the address used for transactions
   */
  updateAccount(address: string): void {
    const config: any = {
      chain: studionet,
      account: address as `0x${string}`,
    };

    this.client = createClient(config);
  }

  /**
   * Get all contests
   */
  async getAllContests(): Promise<Contest[]> {
    try {
      const contests = await this.client.readContract({
        address: this.contractAddress,
        functionName: "get_all_contests",
        args: [],
      });

      // Convert Map structures to plain objects
      const parsed = mapToObject(contests);

      if (Array.isArray(parsed)) {
        return parsed as Contest[];
      }

      return [];
    } catch (error) {
      console.error("Error fetching contests:", error);
      throw new Error("Failed to fetch contests from contract");
    }
  }

  /**
   * Get a specific contest by ID
   */
  async getContest(contestId: number): Promise<Contest> {
    try {
      const contest = await this.client.readContract({
        address: this.contractAddress,
        functionName: "get_contest",
        args: [contestId],
      });

      return mapToObject(contest) as Contest;
    } catch (error) {
      console.error("Error fetching contest:", error);
      throw new Error("Failed to fetch contest from contract");
    }
  }

  /**
   * Get all submissions for a contest
   */
  async getSubmissions(contestId: number): Promise<Submission[]> {
    try {
      console.log("Fetching submissions for contestId:", contestId);
      const submissions = await this.client.readContract({
        address: this.contractAddress,
        functionName: "get_submissions",
        args: [contestId],
      });

      console.log("Raw submissions response:", submissions);
      const parsed = mapToObject(submissions);
      console.log("Parsed submissions:", parsed);

      if (Array.isArray(parsed)) {
        return parsed as Submission[];
      }

      return [];
    } catch (error) {
      console.error("Error fetching submissions:", error);
      throw new Error("Failed to fetch submissions from contract");
    }
  }

  /**
   * Get winners for a contest
   */
  async getWinners(contestId: number): Promise<string[]> {
    try {
      const winners = await this.client.readContract({
        address: this.contractAddress,
        functionName: "get_winners",
        args: [contestId],
      });

      const parsed = mapToObject(winners);

      if (Array.isArray(parsed)) {
        return parsed as string[];
      }

      return [];
    } catch (error) {
      console.error("Error fetching winners:", error);
      throw new Error("Failed to fetch winners from contract");
    }
  }

  /**
   * Get a user's submission for a specific contest
   */
  async getUserSubmission(
    contestId: number,
    userAddress: string
  ): Promise<UserSubmission> {
    try {
      const submission = await this.client.readContract({
        address: this.contractAddress,
        functionName: "get_user_submission",
        args: [contestId, userAddress],
      });

      return mapToObject(submission) as UserSubmission;
    } catch (error) {
      console.error("Error fetching user submission:", error);
      return { has_submitted: false };
    }
  }

  /**
   * Create a new contest
   */
  async createContest(
    platformPattern: string,
    requiredTopic: string,
    rewardDescription: string,
    maxWinners: number,
    deadline: number
  ): Promise<{ receipt: TransactionReceipt; contestId?: number }> {
    try {
      const txHash = await this.client.writeContract({
        address: this.contractAddress,
        functionName: "create_contest",
        args: [
          platformPattern,
          requiredTopic,
          rewardDescription,
          maxWinners,
          deadline,
        ],
        value: BigInt(0),
      });

      const receipt = await this.client.waitForTransactionReceipt({
        hash: txHash,
        status: "ACCEPTED" as any,
        retries: 24,
        interval: 5000,
      });

      // Extract contest ID from receipt if available
      let contestId: number | undefined;
      if (receipt.data?.result !== undefined) {
        contestId = Number(receipt.data.result);
      }

      return { receipt: receipt as TransactionReceipt, contestId };
    } catch (error) {
      console.error("Error creating contest:", error);
      throw new Error("Failed to create contest");
    }
  }

  /**
   * Submit content to a contest
   */
  async submitContent(
    contestId: number,
    contentUrl: string
  ): Promise<{ receipt: TransactionReceipt; result?: SubmitResult }> {
    try {
      const txHash = await this.client.writeContract({
        address: this.contractAddress,
        functionName: "submit_content",
        args: [contestId, contentUrl],
        value: BigInt(0),
      });

      const receipt = await this.client.waitForTransactionReceipt({
        hash: txHash,
        status: "ACCEPTED" as any,
        retries: 60, // Longer wait for LLM validation
        interval: 5000,
      });

      // Extract result from receipt
      let result: SubmitResult | undefined;
      if (receipt.data?.result) {
        result = receipt.data.result as SubmitResult;
      }

      return { receipt: receipt as TransactionReceipt, result };
    } catch (error) {
      console.error("Error submitting content:", error);
      throw new Error("Failed to submit content");
    }
  }

  /**
   * Close a contest (creator only)
   */
  async closeContest(contestId: number): Promise<TransactionReceipt> {
    try {
      const txHash = await this.client.writeContract({
        address: this.contractAddress,
        functionName: "close_contest",
        args: [contestId],
        value: BigInt(0),
      });

      const receipt = await this.client.waitForTransactionReceipt({
        hash: txHash,
        status: "ACCEPTED" as any,
        retries: 24,
        interval: 5000,
      });

      return receipt as TransactionReceipt;
    } catch (error) {
      console.error("Error closing contest:", error);
      throw new Error("Failed to close contest");
    }
  }
}

export default ContentRewards;

File: contracts/content_rewards.py
# { "Depends": "py-genlayer:test" }

import json
from dataclasses import dataclass
from genlayer import *


@allow_storage
@dataclass
class Submission:
    submitter: Address
    content_url: str
    status: str
    rejection_reason: str


@allow_storage
@dataclass
class Contest:
    creator: Address
    platform_pattern: str
    required_topic: str
    reward_description: str
    max_winners: bigint
    deadline: bigint
    accepted_count: bigint
    is_active: bool


class ContentRewards(gl.Contract):
    contests: TreeMap[str, Contest]
    # Simplified: store submissions as a flat map with composite key "contestId:address"
    submissions: TreeMap[str, Submission]
    # Store submission addresses per contest as comma-separated string
    submission_addresses: TreeMap[str, str]
    next_contest_id: bigint

    def __init__(self):
        self.next_contest_id = bigint(0)

    def _validate_content(self, content_url: str, platform_pattern: str, required_topic: str) -> bool:
        # Check platform pattern first (deterministic check)
        if platform_pattern != "*":
            if platform_pattern.lower() not in content_url.lower():
                return False

        def validate_submission() -> str:
            web_data = ""
            fetch_method = ""

            # Method 1: Try mode="text" first (best for readable content)
            try:
                web_data = gl.nondet.web.render(content_url, mode="text")
                if web_data and len(web_data.strip()) > 50:
                    fetch_method = "text content"
            except Exception:
                web_data = ""

            # Method 2: If text failed, try mode="html" for metadata
            if not web_data or len(web_data.strip()) < 50:
                try:
                    html_data = gl.nondet.web.render(content_url, mode="html")
                    if html_data and len(html_data.strip()) > 100:
                        web_data = html_data
                        fetch_method = "HTML metadata"
                except Exception:
                    pass

            # Method 3: If both failed, use URL-based validation
            if not web_data or len(web_data.strip()) < 50:
                fetch_method = "URL analysis only"
                # Extract potential keywords from URL
                web_data = f"URL to analyze: {content_url}"

            # Use LLM to validate topic relevance
            task = f"""
You are evaluating if content is related to a topic for a rewards program.
Data source: {fetch_method}

Required Topic: {required_topic}

Content URL: {content_url}

{"Fetched Content:" if fetch_method != "URL analysis only" else ""}
{web_data[:8000]}

VALIDATION RULES:
1. If the URL contains keywords related to the topic - ACCEPT
2. If the fetched content mentions the topic - ACCEPT
3. Social media posts about the topic ARE VALID even if brief
4. Only REJECT if clearly unrelated or spam
5. When in doubt, ACCEPT

Answer only: YES or NO
"""
            return gl.nondet.exec_prompt(task)

        # Use comparative equivalence - validators compare if their answers are equivalent
        comparison_prompt = """
Two validators evaluated whether content is related to a topic.
Do they reach the same conclusion (both accept OR both reject)?

Validator 1 said: {principal}
Validator 2 said: {other}

Answer only YES if they agree, NO if they disagree.
"""
        result = gl.eq_principle.prompt_comparative(validate_submission, comparison_prompt)

        # Check if result indicates valid content
        result_lower = result.lower().strip()
        return "yes" in result_lower or "valid" in result_lower or "accept" in result_lower

    @gl.public.write
    def create_contest(
        self,
        platform_pattern: str,
        required_topic: str,
        reward_description: str,
        max_winners: int,
        deadline: int
    ) -> int:
        contest_id = int(self.next_contest_id)
        self.next_contest_id = bigint(contest_id + 1)

        contest = Contest(
            creator=gl.message.sender_address,
            platform_pattern=platform_pattern,
            required_topic=required_topic,
            reward_description=reward_description,
            max_winners=bigint(max_winners),
            deadline=bigint(deadline),
            accepted_count=bigint(0),
            is_active=True
        )

        contest_key = str(contest_id)
        self.contests[contest_key] = contest
        # Initialize empty submission addresses for this contest
        self.submission_addresses[contest_key] = ""

        return contest_id

    @gl.public.write
    def submit_content(self, contest_id: int, content_url: str) -> dict:
        contest_key = str(contest_id)

        if contest_key not in self.contests:
            raise Exception("Contest does not exist")

        contest = self.contests[contest_key]

        if not contest.is_active:
            raise Exception("Contest is no longer active")

        if int(contest.deadline) > 0:
            current_time = int(gl.message.timestamp)
            if current_time > int(contest.deadline):
                raise Exception("Contest deadline has passed")

        if int(contest.accepted_count) >= int(contest.max_winners):
            raise Exception("Contest has reached maximum winners")

        sender = gl.message.sender_address
        sender_hex = sender.as_hex

        # Create composite key for submission
        submission_key = f"{contest_key}:{sender_hex}"

        # Check if user already submitted
        if submission_key in self.submissions:
            raise Exception("You have already submitted to this contest")

        # Validate the content
        is_valid = self._validate_content(
            content_url,
            contest.platform_pattern,
            contest.required_topic
        )

        # Determine status based on validation and current accepted count
        if is_valid:
            if int(contest.accepted_count) < int(contest.max_winners):
                status = "accepted"
                contest.accepted_count = bigint(int(contest.accepted_count) + 1)
                rejection_reason = ""

                if int(contest.accepted_count) >= int(contest.max_winners):
                    contest.is_active = False
            else:
                status = "voided"
                rejection_reason = "Contest reached maximum winners during validation"
        else:
            status = "rejected"
            rejection_reason = "Content did not meet topic requirements"

        # Store the submission
        submission = Submission(
            submitter=sender,
            content_url=content_url,
            status=status,
            rejection_reason=rejection_reason
        )

        self.submissions[submission_key] = submission

        # Add sender to submission addresses list
        current_addresses = self.submission_addresses.get(contest_key, "")
        if current_addresses:
            self.submission_addresses[contest_key] = current_addresses + "," + sender_hex
        else:
            self.submission_addresses[contest_key] = sender_hex

        return {
            "status": status,
            "reason": rejection_reason if rejection_reason else "Content accepted"
        }

    @gl.public.write
    def close_contest(self, contest_id: int) -> None:
        contest_key = str(contest_id)

        if contest_key not in self.contests:
            raise Exception("Contest does not exist")

        contest = self.contests[contest_key]

        if gl.message.sender_address != contest.creator:
            raise Exception("Only contest creator can close the contest")

        contest.is_active = False

    @gl.public.view
    def get_contest(self, contest_id: int) -> dict:
        contest_key = str(contest_id)

        if contest_key not in self.contests:
            raise Exception("Contest does not exist")

        contest = self.contests[contest_key]

        return {
            "contest_id": contest_id,
            "creator": contest.creator.as_hex,
            "platform_pattern": contest.platform_pattern,
            "required_topic": contest.required_topic,
            "reward_description": contest.reward_description,
            "max_winners": int(contest.max_winners),
            "deadline": int(contest.deadline),
            "accepted_count": int(contest.accepted_count),
            "is_active": contest.is_active,
            "spots_remaining": int(contest.max_winners) - int(contest.accepted_count)
        }

    @gl.public.view
    def get_all_contests(self) -> list:
        result = []
        for contest_key, contest in self.contests.items():
            result.append({
                "contest_id": int(contest_key),
                "creator": contest.creator.as_hex,
                "platform_pattern": contest.platform_pattern,
                "required_topic": contest.required_topic,
                "reward_description": contest.reward_description,
                "max_winners": int(contest.max_winners),
                "deadline": int(contest.deadline),
                "accepted_count": int(contest.accepted_count),
                "is_active": contest.is_active,
                "spots_remaining": int(contest.max_winners) - int(contest.accepted_count)
            })

        return result

    @gl.public.view
    def get_submissions(self, contest_id: int) -> list:
        contest_key = str(contest_id)

        if contest_key not in self.contests:
            raise Exception("Contest does not exist")

        result = []

        # Get addresses that submitted to this contest
        addresses_str = self.submission_addresses.get(contest_key, "")
        if not addresses_str:
            return result

        addresses = addresses_str.split(",")

        for addr_hex in addresses:
            if not addr_hex:
                continue
            submission_key = f"{contest_key}:{addr_hex}"
            if submission_key in self.submissions:
                sub = self.submissions[submission_key]
                result.append({
                    "submitter": sub.submitter.as_hex,
                    "content_url": sub.content_url,
                    "status": sub.status,
                    "rejection_reason": sub.rejection_reason
                })

        return result

    @gl.public.view
    def get_winners(self, contest_id: int) -> list:
        contest_key = str(contest_id)

        if contest_key not in self.contests:
            raise Exception("Contest does not exist")

        winners = []

        addresses_str = self.submission_addresses.get(contest_key, "")
        if not addresses_str:
            return winners

        addresses = addresses_str.split(",")

        for addr_hex in addresses:
            if not addr_hex:
                continue
            submission_key = f"{contest_key}:{addr_hex}"
            if submission_key in self.submissions:
                sub = self.submissions[submission_key]
                if sub.status == "accepted":
                    winners.append(sub.submitter.as_hex)

        return winners

    @gl.public.view
    def get_user_submission(self, contest_id: int, user_address: str) -> dict:
        contest_key = str(contest_id)

        if contest_key not in self.contests:
            raise Exception("Contest does not exist")

        submission_key = f"{contest_key}:{user_address}"

        if submission_key not in self.submissions:
            return {"has_submitted": False}

        sub = self.submissions[submission_key]
        return {
            "has_submitted": True,
            "content_url": sub.content_url,
            "status": sub.status,
            "rejection_reason": sub.rejection_reason
        }
